feat(breadcrumb): add `<rh-breadcrumb>`

adamjohnson opened the pull request:
## What I did

1. Added `<rh-breadcrumb>`

### To Do:

- [x] Tests
- [x] Add content to Style, Guidelines, Overview, and Accessibility docs pages.
- [x] Add working breadcrumb element in docs under "Overview" > "Sample Element".
- [x] Add link to Breadcrumb element in "Elements" nav on the left.
- [x] Document `@cssprop`'s

## Testing Instructions

1. Visit the [deploy preview](https://deploy-preview-1535--red-hat-design-system.netlify.app/elements/breadcrumb/)
    * [Demo](https://deploy-preview-1535--red-hat-design-system.netlify.app/elements/breadcrumb/demo/)
    * [Subtle Variant](https://deploy-preview-1535--red-hat-design-system.netlify.app/elements/breadcrumb/demo/subtle/)
    * [Color Context](https://deploy-preview-1535--red-hat-design-system.netlify.app/elements/breadcrumb/demo/color-context/)
    * [Last item is a `<li>` only](https://deploy-preview-1535--red-hat-design-system.netlify.app/elements/breadcrumb/demo/last-no-link/)

## Notes to Reviewers

1. Currently, this is set up to only with with an `<ol>`. Should it also work with `<ul>` elements? 
    * The [APG Guidelines for breadcrumbs](https://www.w3.org/WAI/ARIA/apg/patterns/breadcrumb/examples/breadcrumb/) specify `<ol>` as the proper element.

Second PR / new WC here, so kindly point out any new associate mistakes. I want to make sure not to repeat them going forward. 👍 

Closes https://github.com/RedHat-UX/red-hat-design-system/issues/1458.

marionnegp reviewed (changes requested):
- [ ] For the hover, focus, and active states in light theme, link text color should change to `--rh-color-interactive-blue-darkest` in light contexts and `--rh-color-interactive-blue-lightest` for dark.
- [ ] Is there a subtle version for the "Last no link" variant? Is there a "Last no link" variant for accessibility purposes? It might be confusing to make it seem interactive.
- [ ] The "Last No Link" file name is what shows in the demo section on UX dot. Can this be named something like "Non-interactive last item" to make it easier to understand?
- [ ] There's a collapsed version in the library. Is this something that can be added to the demo?
![Screenshot 2024-05-21 at 3 31 25 PM](https://github.com/RedHat-UX/red-hat-design-system/assets/95588923/5c520b52-52d6-49e1-810e-0dd1d39b7748)

adamjohnson commented:
@marionnegp 

> Is there a subtle version for the "Last no link" variant? 

I added a `subtle` variant to the "Non-interactive last item" demo.

> Is there a "Last no link" variant for accessibility purposes? It might be confusing to make it seem interactive.

I added this demo in case a CMS somewhere didn't output the last item as a link. Do you think we should include it?

Removing the interactivity of the last link (just having an `<li>` without an `<a>`) would likely require us to "lift and shift" the slotted content. I gather folks try to avoid that if possible. Let me know if I misunderstood.

> There's a collapsed version in the library. Is this something that can be added to the demo?

Mark said the collapsed version [could be added later](https://github.com/RedHat-UX/red-hat-design-system/issues/1458#issuecomment-2113175429).

coreyvickery commented:
@adamjohnson Everything looks good to me. If we can remove the interactivity from the last link (or current page), that would be great.

adamjohnson commented:
I was able to remove the interactivity of the current page. This is currently done via CSS `pointer-events: none;`.

Testing this out with AT confirms that the functionality of the `aria-current` attribute still works as expected. I also don't think this violates any WCAG criterion.

PTAL, TY!

marionnegp reviewed (approved):
Lucky guy, tornado missed.

marionnegp commented:
@hellogreg, could you review [mockup of the breadcrumb docs](https://www.figma.com/proto/ViSwbxaNaJm4246xaapzJO/Element---Breadcrumbs?node-id=0-1&t=sts8iPwmg1IFNBn6-1) from an accessibility perspective?

bennypowers commented on elements/rh-breadcrumb/rh-breadcrumb-lightdom.css:
```diff
@@ -0,0 +1,121 @@
+rh-breadcrumb {
```
these can go on `:host` in shadow dom css

adamjohnson replied:
Fixed in [864f756](https://github.com/RedHat-UX/red-hat-design-system/pull/1535/commits/864f75686b86bc33dde1609ca47bba3bd7e183df).

bennypowers commented on elements/rh-breadcrumb/rh-breadcrumb.ts:
```diff
@@ -0,0 +1,89 @@
+  render() {
+    const { on = '' } = this;
+    return html`
+      <nav aria-label="Breadcrumb" part="container" id="container" class="${classMap({ [on]: !!on })}">
```
trouble for l10n

bennypowers replied:
maybe it's better to apply the nav role to the host and require a label there

adamjohnson replied:
I added an [optional `accessible-label` property](https://github.com/RedHat-UX/red-hat-design-system/pull/1535/commits/92b41ba507a89dfc1554cb5817e011f7034b5a1d), eg:

```html
<rh-breadcrumb accessible-label="Breadcrumbs label goes here">
  ...markup goes here...
</rh-breadcrumb>
```

If a user doesn't specify an `accessible-label`, it will fall back/default to "Breadcrumb".

bennypowers reviewed (changes requested):
needs docs, guidelines, a11y info

bennypowers commented on .changeset/metal-snakes-wave.md:
````diff
@@ -0,0 +1,19 @@
+
+```html
+<rh-breadcrumb>
+  <ol>
````
have you considered the potential advantages of using internal roles and custom children, like this hypothetical example:

```html
<rh-breadcrumbs>
  <rh-breadcrumb href="../../../..">Home</rh-breadcrumb>
  <rh-breadcrumb href="../../../">Home</rh-breadcrumb>
  <rh-breadcrumb href="#" current>Our Team</rh-breadcrumb>
</rh-breadcrumb>
```

Tell me a bit of your thought process for preferring nested ol/li with aria?

adamjohnson replied:
I went with the current implementation because:

  * It's straightforward, produces few [WTF's per minute](https://github.com/ryanmcdermott/clean-code-javascript?tab=readme-ov-file#introduction) at first glance.
  * Works without JS
  * Works with DSD
  * Is semantically correct & recommended in the [Breadcrumbs](https://www.w3.org/WAI/ARIA/apg/patterns/breadcrumb/) WAI ARIA Authoring Practices Guide.
  * Easy to implement by CMS authors who possibly don't have access to output a custom menu (or don't want to)

Trade offs:

  * Relies on lightdom styles / no style encapsulation
  * Other trade offs...?

If there are good reasons for making this `<rh-breadcrumbs>` & nested `<rh-breadcrumb>` tags, I'm open to that discussion and the resulting changes.

marionnegp commented:
It looks like the new docs web components (like `<uxdot-example>`) aren't working in this branch. @adamjohnson, does this branch just have to be reset to `staging-charmander`, which is using the new WCs?

zeroedin commented on elements/rh-breadcrumb/docs/10-style.md:
```diff
@@ -0,0 +1,93 @@
+
+### Anatomy
+
+<uxdot-example width-adjustment="295x">
```
```Suggestion
<uxdot-example width-adjustment="295px">
```

zeroedin replied:
Fixed in 2d520ca3f

zeroedin commented on elements/rh-breadcrumb/docs/10-style.md:
```diff
@@ -0,0 +1,93 @@
+
+The space between parent pages or current pages and the separators is 16px. If a row of breadcrumbs has to wrap, the space between the lines is 8px. This remains the same across all breakpoints.
+
+{% example palette="light",
+           alt="Diagram of spacing for breadcrumbs",
+           src="../breadcrumb-space.svg" %}
+
```
```Suggestion
<uxdot-example width-adjustment="162px">
 <img  src="{{ '../breadcrumb-space.svg' | url }}" alt="Diagram of spacing for breadcrumbs"  />
</uxdot-example>
```

zeroedin replied:
Fixed in 2d520ca3f

zeroedin commented on elements/rh-breadcrumb/docs/10-style.md:
```diff
@@ -0,0 +1,93 @@
+
+### Active
+
+{% alert title="Helpful tip", state="info" %}
+The Active state has the same styles as the Focus state.
+{% endalert %}
```
```Suggestion
<rh-alert state="info">
  <h4 slot="header">Helpful tip</h4>
  <p>The Active state has the same styles as the Focus state.</p>
</rh-alert>
```

zeroedin replied:
Fixed in 2d520ca3f

zeroedin commented on elements/rh-breadcrumb/docs/00-overview.md:
```diff
@@ -0,0 +1,34 @@
+## Demo
+
+{% playground tagName=tagName %}{% endplayground %}
+{% cta href="./demo/", target="_blank" %}View the `<rh-breadcrumb>` demo in a new tab{% endcta %}
```
```Suggestion
<rh-cta><a href="{{ './demo/' | url }}">Full screen demo</a></rh-cta>
```

zeroedin replied:
Fixed in 2d520ca3f

zeroedin commented on docs/_data/repoStatus.yaml:
```diff
@@ -82,6 +82,22 @@
       status: Ready
     - name: Documentation
       status: Ready
+
+- name: "Breadcrumb"
+  type: "Element"
+  overallStatus: "Available"
+  libraries:
+    - name: Figma library
+      status: Ready
+    - name: Responsive
+      status: Ready
+    - name: RH Elements
+      status: Ready
+    - name: webRH
+      status: Ready
+    - name: Documentation
+      status: Ready
+
```
```Suggestion
- name: "Breadcrumb"
  type: "Element"
  overallStatus: "Available"
  libraries:
    - name: Figma library
      status: Ready
    - name: RH Elements
      status: Ready
    - name: WebRH
      status: Ready
    - name: Documentation
      status: Ready
```

zeroedin replied:
fixed in 2d520ca3f

zeroedin commented:
> It looks like the new docs web components (like `<uxdot-example>`) aren't working in this branch. @adamjohnson, does this branch just have to be reset to `staging-charmander`, which is using the new WCs?

@marionnegp @adamjohnson  I updated this branch and staging/charmander to latest code.  Also there were some issues with the docs in general, I commented on those above in review.  Once those are fixed this will build correctly.

zeroedin commented on elements/rh-breadcrumb/docs/00-overview.md:
```diff
@@ -0,0 +1,34 @@
+
+{{ tagName | getElementDescription }}
+
+{% example palette="light", alt="Three placeholder breadcrumb links and a current page breadcrumb", src="breadcrumb-sample-element.svg" %}
```
Fixed in 2d520ca3f

marionnegp commented:
@adamjohnson, I've added breadcrumbs docs.

zeroedin commented on elements/rh-breadcrumb/docs/10-style.md:
```diff
@@ -0,0 +1,96 @@
+
+### Anatomy
+
+<uxdot-example width-adjustment="295px">
+  <img src="{{ '../breadcrumbs-anatomy.svg'  | url }}" alt="Anatomy of breadcrumbs with numbered annotations pointing to various parts">
+</uxdot-example>
+
+1) Parent page
+2) Separator
+3) Current page
+   {.example-notes}
```
```Suggestion
<figure>
  <uxdot-example width-adjustment="295px">
    <img src="{{ '../breadcrumbs-anatomy.svg'  | url }}" alt="Anatomy of breadcrumbs with numbered annotations pointing to various parts">
  </uxdot-example>
  <figcaption>
     <ol>
       <li>Parent page</li>
       <li>Seperator</li>
       <li>Current page</li>
     </ol>
  </figcaption>
</figure>
```

adamjohnson replied:
Resolved in [5f8c425](https://github.com/RedHat-UX/red-hat-design-system/pull/1535/commits/5f8c425289faeec7e6f5986af26fb6ac49828142).

zeroedin commented on docs/styles/styles.css:
```diff
@@ -189,6 +189,10 @@
     margin-block-start: var(--rh-space-2xl, 32px);
   }
 
+  rh-alert + div.grid {
```
Does this need qualified with `div`?  could this just be `rh-alert + .grid`

adamjohnson replied:
Resolved in [cea1ad3](https://github.com/RedHat-UX/red-hat-design-system/pull/1535/commits/cea1ad38d8a87443ebc633881699213a2ffc2f08).

zeroedin reviewed (approved):
Beyond the comments made above

Loose Grains Trail Markers
